In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import plotly.express  as px
import seaborn as sns

C:\Users\dell\anaconda3\lib\site-packages\scipy\__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.24.0
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


In [2]:
#Reading dataset
df=pd.read_csv("D:\AIF Python\Data sets\Project Data set\carvana.csv")

In [3]:
#It gives first five rows
df.head()

Name   Year  Miles  Price
0   Chevrolet Trax   2018  41946  16990
1      GMC Terrain   2020  45328  23990
2    Jeep Wrangler   2012  81068  21590
3    Jeep Renegade   2019  35372  21590
4            BMW X  20173  68992  22990

In [4]:
#It gives last 5 rows
df.tail()

Name  Year  Miles  Price
21995               Honda Accord  2014  98085  17590
21996              Dodge Charger  2016  62616  23990
21997                Ford Taurus  2013  50345  16990
21998   Toyota Corolla Hatchback  2020  27945  23990
21999                Honda Civic  2012  70676  14990

In [5]:
#this function gives us count, Minimum and Maximum
df.describe()

Year          Miles          Price
count  2.200000e+04   22000.000000   22000.000000
mean   1.946648e+05   54445.697318   20707.285136
std    1.630370e+06   25685.933926    6535.580304
min    2.009000e+03      53.000000   10990.000000
25%    2.014000e+03   35512.000000   16590.000000
50%    2.017000e+03   51890.000000   19590.000000
75%    2.020000e+03   73191.000000   22990.000000
max    2.022350e+07  120167.000000  102990.000000

In [6]:
#This function gives us null values count, datatypes of columns, column names
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22000 entries, 0 to 21999
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Name    22000 non-null  object
 1   Year    22000 non-null  int64 
 2   Miles   22000 non-null  int64 
 3   Price   22000 non-null  int64 
dtypes: int64(3), object(1)
memory usage: 687.6+ KB


In [7]:
#In Year column some of the values are more than four digits so we are correcting them using slicing
df["Year"] = df["Year"].astype(str).str[0:4].astype(int)

In [8]:
#After correcting Year column we can see that all years now in 4 digit 
df.head()

Name  Year  Miles  Price
0   Chevrolet Trax  2018  41946  16990
1      GMC Terrain  2020  45328  23990
2    Jeep Wrangler  2012  81068  21590
3    Jeep Renegade  2019  35372  21590
4            BMW X  2017  68992  22990

In [9]:
#delete are all the duplicate entries and keeps one
df = df.drop_duplicates() 

In [10]:
#after dropping duplicates now many entries we have
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9318 entries, 0 to 21999
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Name    9318 non-null   object
 1   Year    9318 non-null   int32 
 2   Miles   9318 non-null   int64 
 3   Price   9318 non-null   int64 
dtypes: int32(1), int64(2), object(1)
memory usage: 327.6+ KB


Before dropping we have 22000 entries and after dropping now we have 9318 entry left

In [11]:
#The strip() method removes any leading (spaces at the beginning) and trailing (spaces at the end) characters (space is the default leading character to remove)
df["Name"]= df["Name"].str.strip()

In [12]:
#searching for how many years data we have
df["Year"].nunique()

15

In [13]:
df.groupby("Year").count().sort_values("Year")

Name  Miles  Price
Year                    
2009    52     52     52
2010   133    133    133
2011   202    202    202
2012   440    440    440
2013   638    638    638
2014   810    810    810
2015   977    977    977
2016  1077   1077   1077
2017  1075   1075   1075
2018   968    968    968
2019   941    941    941
2020   907    907    907
2021   817    817    817
2022   278    278    278
2023     3      3      3

From the above data we can conclude that we have data from 2009 to 2023

In [14]:
df.groupby("Year").count().sort_values("Name")

Name  Miles  Price
Year                    
2023     3      3      3
2009    52     52     52
2010   133    133    133
2011   202    202    202
2022   278    278    278
2012   440    440    440
2013   638    638    638
2014   810    810    810
2021   817    817    817
2020   907    907    907
2019   941    941    941
2018   968    968    968
2015   977    977    977
2017  1075   1075   1075
2016  1077   1077   1077

From the above data we can conclude that Maximum car's available for resale are buyed in 2016.

In [15]:
df.loc[df["Name"].str.contains("BMW")]

Name  Year  Miles  Price
4            BMW X  2017  68992  22990
105    BMW  Series  2011  64514  22990
129    BMW  Series  2013  77600  18990
353          BMW i  2015  44100  20590
420          BMW i  2014  33951  18990
...            ...   ...    ...    ...
18099  BMW  Series  2014  57799  21590
20099  BMW  Series  2015  63832  19990
20694        BMW i  2017  60954  21990
20747  BMW  Series  2014  80071  18990
21765  BMW  Series  2010  71435  17990

[240 rows x 4 columns]

From the above data we conclude that we have 240 car's availabe for resale manufactured by BMW company

In [16]:
df.groupby("Name").count().sort_values("Year", ascending = False)

Year  Miles  Price
Name                                                    
Toyota Corolla                         269    269    269
Nissan Rogue                           242    242    242
Honda Civic                            210    210    210
Hyundai Elantra                        195    195    195
Nissan Altima                          188    188    188
...                                    ...    ...    ...
Toyota                                   1      1      1
Chevrolet Silverado  HD Regular Cab      1      1      1
Mercedes-Benz Mercedes-AMG GLC           1      1      1
Mercedes-Benz Mercedes-AMG GLE Coupe     1      1      1
smart fortwo electric drive cabrio       1      1      1

[413 rows x 3 columns]

We have total 416 various models available for resale "Toyota Corolla" model have highest number of cars available for resale.

In [17]:
#created new dataframe where all the car's are made in 2015
df_2015 = df.loc[df["Year"]==2015] 

In [18]:
df_2015.sort_values("Miles", ascending = False)

Name  Year   Miles  Price
3157             Ford Fiesta  2015  106396  11990
11901          Ford Explorer  2015  105432  17590
10330           Jeep Compass  2015  105122  14590
3911          Toyota Corolla  2015  104984  15990
7837                Scion tC  2015  104712  15590
...                      ...   ...     ...    ...
1879      MINI Hardtop  Door  2015   10989  22990
10639        Hyundai Elantra  2015    6849  17990
4120            Ford Mustang  2015    5035  34590
5592            Ford Mustang  2015    1431  38590
6208   Hyundai Genesis Coupe  2015    1289  28990

[977 rows x 4 columns]

We have total 977 cars available for resale from year 2015, The most driven car is Ford Fiesta which is 106396 Miles driven and price is 11990

In [19]:
df_2015.sort_values("Miles")

Name  Year   Miles  Price
6208   Hyundai Genesis Coupe  2015    1289  28990
5592            Ford Mustang  2015    1431  38590
4120            Ford Mustang  2015    5035  34590
10639        Hyundai Elantra  2015    6849  17990
1879      MINI Hardtop  Door  2015   10989  22990
...                      ...   ...     ...    ...
7837                Scion tC  2015  104712  15590
3911          Toyota Corolla  2015  104984  15990
10330           Jeep Compass  2015  105122  14590
11901          Ford Explorer  2015  105432  17590
3157             Ford Fiesta  2015  106396  11990

[977 rows x 4 columns]

The less driven car is 'Hyundai Genesis Coupe'  which is only 1289 Miles driven and price is 28990

In [20]:
df_2015.sort_values("Price")

Name  Year   Miles  Price
3157         Ford Fiesta  2015  106396  11990
1197        smart fortwo  2015   82396  11990
6370         Ford Fiesta  2015  100592  11990
9522         Ford Fiesta  2015   98161  12590
1348         Ford Fiesta  2015   99415  12590
...                  ...   ...     ...    ...
4334               BMW Z  2015   17597  40990
4504  Chevrolet Corvette  2015   19307  47990
5480    Dodge Challenger  2015   51184  50990
4958  Chevrolet Corvette  2015   11617  51590
4127  Chevrolet Corvette  2015   12748  66990

[977 rows x 4 columns]

From year 2015 'Chevrolet Corvette' is the most expensive car available for resale which is driven 12748 miles and 'Ford Fiesta' is the less expensive car which is driven 106396 miles.

In [25]:
plt.figure(figsize=(15,7))
fig = px.bar(df_2015.sort_values("Price").tail(10), x= "Name" , y= "Price",
       title= "10 Most expensive cars made in 2015", color= "Name")
fig.show()

<Figure size 1080x504 with 0 Axes>

In top 10 expensive cars which are  made in 2015, 4 are from 'Chevrolet' company and 3 are from 'Ford' company.